stdlib/Pkg/src/API.jl

```diff
@@ -445,52 +445,33 @@ precompile() = precompile(Context())
 function precompile(ctx::Context)
     printpkgstyle(ctx, :Precompiling, "project...")
 
-    pkgids = [Base.PkgId(UUID(uuid), name) for (name, uuid) in ctx.env.project["deps"] if !(UUID(uuid) in  keys(ctx.stdlibs))]
-    if ctx.env.pkg !== nothing && isfile( joinpath( dirname(ctx.env.project_file), "src", ctx.env.pkg.name * ".jl"))
+    pkgids = [Base.PkgId(UUID(uuid), name) for (name, uuid) in ctx.env.project["deps"] if !(UUID(uuid) in keys(ctx.stdlibs))]
+    if ctx.env.pkg !== nothing && isfile( joinpath( dirname(ctx.env.project_file), "src", ctx.env.pkg.name * ".jl") )
         push!(pkgids, Base.PkgId(ctx.env.pkg.uuid, ctx.env.pkg.name))
     end
 
-    needs_to_be_precompiled = String[]
+    # TODO: since we are a complete list, but not topologically sorted, handling of recursion will be completely at random
     for pkg in pkgids
         paths = Base.find_all_in_cache_path(pkg)
         sourcepath = Base.locate_package(pkg)
         if sourcepath == nothing
+            # XXX: this isn't supposed to be fatal
             cmderror("couldn't find path to $(pkg.name) when trying to precompilie project")
         end
-        found_matching_precompile = false
+        stale = true
         for path_to_try in paths::Vector{String}
             staledeps = Base.stale_cachefile(sourcepath, path_to_try)
-            if !(staledeps isa Bool)
-                found_matching_precompile = true
-            end
+            staledeps === true && continue
+            # TODO: else, this returns a list of packages that may be loaded to make this valid (the topological list)
+            stale = false
+            break
         end
-        if !found_matching_precompile
-            # Don't bother attempting to precompile packages that appear to contain `__precompile__(false)`
-            source = read(sourcepath, String)
-            if !occursin(r"__precompile__\(false\)", source)
-                push!(needs_to_be_precompiled, pkg.name)
-            end
+        if stale
+            printpkgstyle(ctx, :Precompiling, pkg)
```

Comment:
`pkg` is a `PkgId` but this function expects the third argument to be a string, which is why the tests are failing.